### DISCUSSION TITLE
Strange nameof() expression behavior over alias-qualified references

### DISCUSSION THREAD

**Discussion by controlflow** [dotnet/csharplang#6615] [Language Ideas]
``` c#
using System;

public class Foo {
    public void Bar() {
        // Error CS8083: An alias-qualified name is not an expression
        var a = nameof(global::System);
        // ok, "Collections"
        var b = nameof(global::System.Collections);
    }
}
```

**Comment by jveselka** [comment]
Just ran into this issue. @gafter Why is it not a bug?

**Comment by gafter** [comment]
This is not a bug because [the spec](https://github.com/dotnet/csharplang/blob/master/spec/expressions.md#nameof-expressions) says that the argument to `nameof` is syntactically a expression. A name of the form `A::B` is not an expression.

**Comment by mgravell** [comment]
when `B` is a type in the namespace root, `alias::B` (usually `global::B`, but not necessarily) is *absolutely* a valid expression:

- `new alias::B()` works
- `typeof(alias::B)` works
- `alias::B.SomeStaticMethod()` works
- but `nameof(alias::B)` does not work

I acknowledge that it isn't a valid `nameof_expression` - but that is arguably an error in the specification of `nameof_expression`.

**Comment by CyrusNajmabadi** [comment]
Moving to discussion on csharplang as this is a request for the language to support this.

**Comment by Rekkonnect** [comment]
`nameof` has some running issues and a large margin to be given some love, some things that would otherwise seem possible aren't like using open generic types as the result string ignores the generic type arguments.

**Reply by jnm2** [reply]
For open types in nameof, see https://github.com/dotnet/csharplang/discussions/702#discussioncomment-4022086
